src/cockpit/bridge.py

```diff
@@ -111,9 +111,19 @@ def get_os_release():
                 logger.warning("Neither /etc/os-release nor /usr/lib/os-release exists")
                 return {}
 
-        with file:
-            lexer = shlex.shlex(file, posix=True, punctuation_chars=True)
-            return dict(token.split('=', 1) for token in lexer)
```

Comment:
Why not keep the current approach but unroll the loop?

Something like

```python
os_release = {}
with file:
    for token in shlex.shlex(file, posix=True, punctuation_chars=True):
        with contextlib.suppress(ValueError):
            key, value = token.split('=', 1)
            os_release[key] = value
    return os_release
```

Reply:
The strict "token by token" based approach has rather poor error recovery. The os-release format is line based, but tokens are not. If you e.g. have a line `X=a:b`, that will be parsed as three tokens `X=a`, `:`, and `b`. This is not at all what shells do, `X=a:b` works just fine. Anyway, it still eventually recovers when it finds the next valid `key=value`, but it will have an actively wrong value for `X` in that example, instead of rejecting it flat out.

For this diff:

```diff
--- src/cockpit/bridge.py
+++ src/cockpit/bridge.py
@@ -112,17 +112,13 @@ class Bridge(Router, PackagesListener):
                 return {}
 
         os_release = {}
-        for line in file.readlines():
-            line = line.strip()
-            if not line or line.startswith('#'):
-                continue
-            try:
-                k, v = line.split('=')
-                (v_parsed, ) = shlex.split(v)  # expect exactly one token
-            except ValueError:
-                logger.warning('Ignoring invalid line in os-release: %r', line)
-                continue
-            os_release[k] = v_parsed
+        with file:
+            for token in shlex.shlex(file, posix=True, punctuation_chars=True):
+                try:
+                    key, value = token.split('=', 1)
+                except ValueError:
+                    logger.warning('Ignoring invalid assignment in os-release: %r', token)
+                os_release[key] = value
         return os_release
 
     def do_init(self, message: Dict[str, object]) -> None:
diff --git test/pytest/test_bridge.py test/pytest/test_bridge.py
index 612f7a025..4013b2468 100644
--- test/pytest/test_bridge.py
+++ test/pytest/test_bridge.py
@@ -566,7 +566,7 @@ async def test_channel(channeltype, tmp_path):
     ),
     # ignore ungrammatical lines
     (
-        'A=a\nNOVALUE\nDOUBLEEQ=a=b\nB=b',
+        'A=a\nNOVALUE\nNOEQ val\nB=b',
         {'A': 'a', 'B': 'b'}
     ),
     # invalid values; anything outside [A-Za-z0-9] must be quoted; but our parser is more liberal

```

you get

```
WARNING  cockpit.bridge:bridge.py:120 Ignoring invalid assignment in os-release: ':'
WARNING  cockpit.bridge:bridge.py:120 Ignoring invalid assignment in os-release: 'b'
WARNING  cockpit.bridge:bridge.py:120 Ignoring invalid assignment in os-release: 'b'

E           AssertionError: assert {'X': 'a', 'Y': 'a', 'Z': 'a-b', 'V': 'a_b'} == {'X': 'a:b', 'Z': 'a-b', 'V': 'a_b'}
E             Common items:
E             {'V': 'a_b', 'Z': 'a-b'}
E             Differing items:
E             {'X': 'a'} != {'X': 'a:b'}
E             Left contains 1 more item:
E             {'Y': 'a'}
E             Full diff:
E             - {'V': 'a_b', 'X': 'a:b', 'Z': 'a-b'}
E             ?                     --
E             + {'V': 'a_b', 'X': 'a', 'Y': 'a', 'Z': 'a-b'}
E             ?                       ++++++++++
```

I'm not sure if it's worth keeping the simple tokenizer approach for this rather poor behavior. @allisonkarlitskaya WDYT?

Reply:
The immediate problem at hand could be solved by adding `:` to `wordchars` like:

```python
  lexer.wordchars += ':'
```

Apparently the default values there are alnums and `_`.  Maybe we could add some other things as well?  And maybe also reconsider our use of `punctuation_chars`.

```python
   lexer.whitespace_split = True
```

that will solve this.

I agree that I'm somewhat less enamored of `shlex` than I used to be, though...

It's just kinda frustrating to see what used to be a nice tidy two-liner get blown up into a larger thing...

Reply:
OK, here's my best attempt to do the processing along your lines:

```python

def parse_os_release(content: str) -> Iterable[Tuple[str, str]]:
    for line in content.splitlines():
        with contextlib.suppress(ValueError):
            match = re.fullmatch(r'(\w+)=(.*)', line)
            if match is not None:
                (single_value,) = shlex.split(match.group(2))
                yield match.group(1), single_value
                continue
        if not re.fullmatch(r'\s*(#.*)?', line):
            print('bzzt invalid line', line)


def do_the_thing():
    return dict(parse_os_release(Path('/etc/os-release').read_text()))
```

Reply:
At this point I think we've gone around enough times that (a) you've convinced me that a line-based approach is probably better (and indeed, the documentation is structured this way) and (b) I trust whatever you decide.

Reply:
Why did you add `.strip()` and duplicate the `logger.warning()` line?  It should already have worked that if we didn't yield the value, we'd have fallen through and displayed the warning...

Reply:
str.splitlines() removes `\n`, but file.readlines() doesn't. They could be added to the regexps, but I'm not sure about funny things like `\r`. The duplicated logger.warning() could have been avoided for sure, but it'd still look much harder to understand imho.

Reply:
Sure.  Happy to go with your solution :)